### PR TITLE
Refactor the label-related methods

### PR DESCRIPTION
**Pull request by vladmos**
Create a special package `labels` with all label-related utility functions. It will be possible to load it from anywhere without circular dependencies, also that will unify the way labels are parsed and printed across the codebase.

### CODE REVIEW

**Review by laurentlb — APPROVED** [review]
lgtm

File: labels/labels.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	"bytes"
+	"path"
+	"strings"
+)
+
+// Label represents a Bazel target label.
+type Label struct {
+	Repository string
+	Package    string
+	Target     string
```
[the remark this answers is not in the document]
Comment (vladmos) [reply labels/labels.go]: Done.

File: labels/labels.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+func (l Label) Format() string {
```
[the remark this answers is not in the document]
Comment (vladmos) [reply labels/labels.go:36]: Thanks, I decided to write the comments later when the exported methods and their signatures are finalized, and of course forgot to do that.

File: labels/labels.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+		label.Target = target
+		label.Package = ""
+	} else {
+		// "//absolute/pkg" -> "absolute/pkg", "pkg"
+		label.Target = path.Base(parts[0])
+	}
+	return label
+}
+
+func ParseRelativeLabel(target, pkg string) Label {
```
[the remark this answers is not in the document]
Comment (vladmos) [reply labels/labels.go]: Renamed `target` to `input` and commented.

**Comment by vladmos** [comment]
I've also replaced another custom method that parses relative labels with the new from the `labels` module.

File: labels/labels.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+		// Maybe it's not a label at all, e.g. a filename.
+		return input
+	}
+	label := ParseLabel(input)
+	return label.FormatRelative(pkg)
+}
+
+// Equal returns true if label1 and label2 are equal. The function
+// takes care of the optional ":" prefix and differences between long-form
+// labels and local labels.
```
Comment (laurentlb) [inline labels/labels.go:118]: `local labels (relative to pkg).`

**Review by laurentlb — APPROVED** [review]

File: labels/labels.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+func ParseRelativeLabel(input, pkg string) Label {
+	if !strings.HasPrefix(input, "@") && !strings.HasPrefix(input, "//") {
+		return Label{Package: pkg, Target: strings.TrimLeft(input, ":")}
+	}
+	return ParseLabel(input)
+}
+
+// ShortenLabel rewrites labels to use the canonical form (the form
+// recommended by build-style).
+// "//foo/bar:bar" => "//foo/bar", or ":bar" when possible.
```
[the remark this answers is not in the document]
Comment (vladmos) [reply labels/labels.go]: Done.

File: labels/labels.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	b.WriteString(l.Package)
+	if l.Target != path.Base(l.Package) {
+		b.WriteString(":")
+		b.WriteString(l.Target)
+	}
+	return b.String()
+}
+
+// FormatRelative returns a string representation of a label relative to `pkg`
+// (relative label if it represent a target in the same package, absolute otherwise)
```
[the remark this answers is not in the document]
Comment (vladmos) [reply labels/labels.go]: Done.

File: labels/labels.go (hunk 4 above)
Comment (vladmos) [reply labels/labels.go]: Done.

File: labels/labels.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+		return input
+	}
+	label := ParseLabel(input)
+	return label.FormatRelative(pkg)
+}
+
+// Equal returns true if label1 and label2 are equal. The function
+// takes care of the optional ":" prefix and differences between long-form
+// labels and local labels.
+func Equal(label1, label2, pkg string) bool {
```
[the remark this answers is not in the document]
Comment (vladmos) [reply labels/labels.go:119]: I'm not sure, it's not a method of a label but a function that compares two labels: `labels.Equal(l1, l2, pkg)`. It looks wrong to me if it's called `Equals` or `IsEqual`. E.g. see https://golang.org/pkg/reflect/#DeepEqual

**Review by laurentlb — APPROVED** [review]
lgtm

**Pull request merged**